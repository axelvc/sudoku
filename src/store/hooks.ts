import { useDispatch, useSelector, useStore, TypedUseSelectorHook } from 'react-redux'
import store, { AppDispatch, RootState } from './store'

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
export const useAppStore: () => typeof store = useStore
