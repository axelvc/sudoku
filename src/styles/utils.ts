import { css } from 'styled-components'

export const breakpoints = {
  md: '768px',
}

export const roundedGrid = (rows: number, cols: number): ReturnType<typeof css> => css`
  display: grid;
  grid-template: ${p => `repeat(${rows}, 1fr) / repeat(${cols}, 1fr) `};
  gap: 0.25rem;
  overflow: hidden;
  border-radius: 0.5rem;
`

export const fancyBox = css`
  padding: 1.5rem;
  background: var(--color-gray-100);
  border-radius: 1rem;
`
