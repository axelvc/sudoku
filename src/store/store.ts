import { Action, configureStore, ThunkAction } from '@reduxjs/toolkit'

import controlsSlice from '../controls/controlsSlice'
import sudokuSlice from '../sudoku/sudokuSlice'

const store = configureStore({
  reducer: {
    controls: controlsSlice,
    sudoku: sudokuSlice,
  },
})

export default store

export type AppDispatch = typeof store.dispatch
export type RootState = ReturnType<typeof store.getState>
// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>
