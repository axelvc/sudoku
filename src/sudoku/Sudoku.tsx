/* eslint-disable @typescript-eslint/no-unused-vars */
import { FC, useEffect, useState } from 'react'
import { useAppDispatch, useAppSelector, useAppStore } from '../store/hooks'
import { difficulties, fetchSudoku } from './sudokuService'
import { fillBox, setSudoku, Coords } from './sudokuSlice'

import DropDown from '../common/DropDown/DropDown'
import Timer from './Timer'
import { Box, Header, Mark, SudokuGrid } from './Sudoku.style'

const Sudoku: FC = () => {
  const [difficulty, setDifficulty] = useState('easy')
  const puzzle = useAppSelector(state => state.sudoku.puzzle)
  const dispatch = useAppDispatch()
  const store = useAppStore()

  useEffect(() => {
    fetchSudoku(difficulty)
      .then(res => dispatch(setSudoku(res)))
      .catch(() => {
        // TODO: handle error
      })
  }, [difficulty])

  function fillBoxWithControlValue(coords: Coords): void {
    const { fillValue: value, marksEnabled } = store.getState().controls

    if (value === null) return

    dispatch(fillBox({ ...coords, value, isMark: marksEnabled }))
  }

  return (
    <section>
      <Header>
        <Timer />
        <DropDown
          options={difficulties}
          selected={difficulty}
          onChange={value => setDifficulty(value)}
        />
      </Header>

      <SudokuGrid>
        {puzzle.map((arr, row) =>
          arr.map((box, col) => (
            <Box
              key={`${row}${col}`}
              data-testid={`box-${row}-${col}`}
              hasMarks={box.marks.length >= 1}
              error={box.errors > 0}
              disabled={box.blocked}
              onClick={() => fillBoxWithControlValue({ row, col })}
            >
              {box.value ? (
                <span>{box.value}</span>
              ) : (
                box.marks.map(mark => (
                  <Mark key={mark} mark={mark}>
                    {mark}
                  </Mark>
                ))
              )}
            </Box>
          )),
        )}
      </SudokuGrid>
    </section>
  )
}

export default Sudoku
