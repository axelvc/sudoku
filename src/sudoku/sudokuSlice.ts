import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { Sudoku, ResponseType } from './sudokuService'
import checkCollisions from './utils/checkCollisions'

export interface Coords {
  row: number
  col: number
}

export interface FillData extends Coords {
  value: number
  isMark?: boolean
}

export interface BoxData {
  value: number
  marks: number[]
  errors: number
  blocked: boolean
}

export interface SudokuState {
  solution: Sudoku
  puzzle: BoxData[][]
}

function parsePuzzle(puzzle: Sudoku): BoxData[][] {
  return puzzle.map(row =>
    row.map(value => ({
      value,
      marks: [],
      errors: 0,
      blocked: value !== 0,
    })),
  )
}

const initialState: SudokuState = {
  solution: Array(9).fill(Array(9).fill(0)),
  puzzle: parsePuzzle(Array(9).fill(Array(9).fill(0))),
}

const sudokuSlice = createSlice({
  name: 'sudoku',
  initialState,
  reducers: {
    setSudoku(state, { payload: { puzzle, solution } }: PayloadAction<ResponseType>) {
      state.solution = solution
      state.puzzle = parsePuzzle(puzzle)
    },
    fillBox(state, { payload: { row, col, value, isMark } }: PayloadAction<FillData>) {
      const box = state.puzzle[row][col]
      const oldValue = box.value

      if (box.blocked) return

      if (value === 0) {
        box.value = 0
        box.marks = []
      } else if (isMark) {
        const index = box.marks.indexOf(value)

        if (index === -1) {
          box.marks.push(value)
        } else {
          box.marks.splice(index, 1)
        }

        box.value = 0
      } else {
        box.value = box.value === value ? 0 : value
        box.marks = []
      }

      if (oldValue !== box.value) {
        checkCollisions(state, { row, col }, oldValue)
      }
    },
    validateSudoku(state) {
      const { puzzle, solution } = state

      puzzle.forEach((row, rI) =>
        row.forEach((box, cI) => {
          const { value, errors } = box
          const rightValue = solution[rI][cI]

          if (value && !errors && value !== rightValue) {
            box.errors++
          }
        }),
      )
    },
  },
})

export const { setSudoku, fillBox, validateSudoku } = sudokuSlice.actions

export default sudokuSlice.reducer
