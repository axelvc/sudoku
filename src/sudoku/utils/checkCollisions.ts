import { Coords, BoxData, SudokuState } from '../sudokuSlice'

function getRowSiblings(puzzle: BoxData[][], { row, col }: Coords): BoxData[] {
  return puzzle[row].filter((box, i) => box.value && i !== col)
}

function getColSiblings(puzzle: BoxData[][], { row, col }: Coords): BoxData[] {
  return puzzle.reduce((siblings, r, i) => {
    const box = r[col]

    if (box.value && i !== row) {
      siblings.push(box)
    }

    return siblings
  }, [])
}

function getBlockSiblings(puzzle: BoxData[][], { row, col }: Coords): BoxData[] {
  const blockRow = Math.floor(row / 3)
  const blockCol = Math.floor(col / 3)

  return puzzle.reduce((siblings, r, rI) => {
    r.forEach((box, cI) => {
      if (
        box.value &&
        rI !== row &&
        cI !== col &&
        blockRow === Math.floor(rI / 3) &&
        blockCol === Math.floor(cI / 3)
      ) {
        siblings.push(box)
      }
    })

    return siblings
  }, [])
}

export default function checkCollisions(
  state: SudokuState,
  { row, col }: Coords,
  oldValue: number,
): void {
  const { puzzle } = state
  const box = puzzle[row][col]
  const siblings = [
    ...getRowSiblings(puzzle, { row, col }),
    ...getColSiblings(puzzle, { row, col }),
    ...getBlockSiblings(puzzle, { row, col }),
  ]

  if (box.errors) {
    box.errors = 0

    siblings.forEach(sibling => {
      if (sibling.value === oldValue) {
        sibling.errors--
      }
    })
  }

  siblings.forEach(sibling => {
    if (sibling.value === box.value) {
      sibling.errors++
      box.errors++
    }
  })
}
