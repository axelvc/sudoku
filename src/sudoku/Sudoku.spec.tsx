import { RootStore } from '../store/store'
import { regex, render, screen, userEvent } from '../test/utils'

import Sudoku from './Sudoku'
import { Coords, fillBox } from './sudokuSlice'

jest.mock('./sudokuService')

let store: RootStore
beforeEach(() => {
  const r = render(<Sudoku />)

  store = r.store
})

describe('Sudoku component', () => {
  describe('collision feedback', () => {
    const cases: Array<[string, Coords]> = [
      ['row', { row: 0, col: 1 }],
      ['col', { row: 1, col: 0 }],
      ['block', { row: 1, col: 1 }],
    ]

    it.each(cases)('should add error styles by collision on %s', (_, coords) => {
      const $box = screen.getByTestId('box-0-0')

      store.dispatch(fillBox({ row: 0, col: 0, value: 1 }))
      store.dispatch(fillBox({ ...coords, value: 1 }))

      expect($box).toMatchSnapshot()
    })
  })

  describe('functionality with controls', () => {
    describe('numpad', () => {
      const numpadNumbers = [1, 2, 3, 4, 5, 6, 7, 8, 9].map(String)

      it('should fill box value', () => {
        const $box = screen.getByTestId('box-0-0')

        numpadNumbers.forEach(n => {
          const $control = screen.getByTestId(`numpad-${n}`)

          userEvent.click($control)
          userEvent.click($box)

          expect($box).toHaveTextContent(n)
        })
      })

      it('should add mark value', () => {
        const $box = screen.getByTestId('box-0-0')
        const $marks = screen.getByRole('button', { name: regex('marks') })

        userEvent.click($marks)

        numpadNumbers.forEach(n => {
          const $control = screen.getByTestId(`numpad-${n}`)

          userEvent.click($control)
          userEvent.click($box)
        })

        expect($box).toHaveTextContent(numpadNumbers.join(''))
      })

      it('should add selected styles on click', () => {
        const $box = screen.getByTestId('box-0-0')
        const $control = screen.getByTestId('numpad-1')

        userEvent.click($control)
        userEvent.click($box)

        expect($box).toMatchSnapshot()
      })
    })

    describe('utils', () => {
      it('should clean box on click with "erase" control enabled', () => {
        const $box = screen.getByTestId('box-0-0')
        const $erase = screen.getByRole('button', { name: regex('erase') })
        const $numpadControl = screen.getByTestId('numpad-1')

        userEvent.click($numpadControl)
        userEvent.click($box)
        userEvent.click($erase)
        userEvent.click($box)

        expect($box).toHaveTextContent('')
      })

      it('should add error styles on click "validate" control', () => {
        const $validate = screen.getByRole('button', { name: regex('validate') })
        const $box = screen.getByTestId('box-0-0')
        const $numpad1 = screen.getByTestId('numpad-1')

        userEvent.click($numpad1)
        userEvent.click($box)
        userEvent.click($validate)

        expect($box).toHaveStyleRule('--color', 'var(--color-red-500)')
        expect($box).toHaveStyleRule('--background', 'var(--color-red-100)')
        expect($box).toMatchSnapshot()
      })

      it('should restore value on click "undo" control after set value', () => {
        const $undo = screen.getByRole('button', { name: regex('undo') })
        const $box = screen.getByTestId('box-0-0')
        const $numpad1 = screen.getByTestId('numpad-1')

        userEvent.click($numpad1)
        userEvent.click($box)
        userEvent.click($undo)

        expect($box).toHaveTextContent('')
      })

      it('should restore all changed values on click "reset" control', () => {
        const $reset = screen.getByRole('button', { name: regex('reset') })
        const $box1 = screen.getByTestId('box-0-0')
        const $box2 = screen.getByTestId('box-0-1')
        const $numpad1 = screen.getByTestId('numpad-1')

        userEvent.click($numpad1)
        userEvent.click($box1)
        userEvent.click($box2)

        userEvent.click($reset)
        expect($box1).toHaveTextContent('')
        expect($box2).toHaveTextContent('')
      })
    })
  })
})
