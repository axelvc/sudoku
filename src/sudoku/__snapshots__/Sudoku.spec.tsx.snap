// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sudoku component collision feedback should add error styles by collision on block 1`] = `
.c0 {
  display: grid;
  grid-auto-flow: column;
  gap: 0.25rem;
  place-items: center;
  min-height: 2rem;
  padding: 0 0.75rem;
  font-weight: 600;
  color: var(--color);
  text-transform: capitalize;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: var(--background);
  border-radius: 0.5rem;
  outline-color: var(--color);
  -webkit-transition: 100ms ease-out;
  transition: 100ms ease-out;
  -webkit-transition-property: background,color;
  transition-property: background,color;
  --color: var(--color-blue-400);
  --color-hover: var(--color);
  --color-active: var(--color-gray-100);
  --background: var(--color-blue-100);
  --background-hover: var(--color-blue-200);
  --background-active: var(--color-blue-400);
}

.c0:hover,
.c0:focus-visible {
  color: var(--color-hover);
  background: var(--background-hover);
}

.c0:active {
  color: var(--color-active);
  background: var(--background-active);
}

.c1[aria-pressed='true'] {
  --color: var(--color-gray-100);
  --background: var(--color-blue-400);
  --background-hover: var(--color-blue-500);
  --background-active: var(--color-blue-600);
}

.c1[aria-pressed='true']:focus {
  outline-color: var(--color-gray-900);
}

.c2 {
  position: relative;
  aspect-ratio: 1;
  display: grid;
  place-items: center;
  padding: 0;
  font-weight: 700;
  border-radius: 0.2rem;
  --color: var(--color-gray-900);
  --color-active: var(--color-gray-900);
  --background: var(--color-gray-200);
  --background-hover: var(--color-gray-300);
  --background-active: var(--color-gray-400);
  --disabled-line: var(--color-gray-300);
  --background: var(--color-gray-300);
  --background-hover: var(--color-gray-400);
  --background-active: var(--color-gray-500);
  --disabled-line: var(--color-gray-500);
}

.c2:disabled {
  background-image: linear-gradient( -45deg, var(--background) 15%, var(--disabled-line) 15% 25%, var(--background) 25% 35%, var(--disabled-line) 35% 45%, var(--background) 45% 55%, var(--disabled-line) 55% 65%, var(--background) 65% 75%, var(--disabled-line) 75% 85%, var(--background) 85% );
  --color: var(--color-gray-600);
}

.c2[aria-pressed='true'] {
  --color: var(--color-green-400);
  --color-active: var(--color-gray-100);
  --background: var(--color-green-100);
  --background-hover: var(--color-green-200);
  --background-active: var(--color-green-400);
  --disabled-line: var(--color-green-200);
}

.c2[aria-invalid='true'] {
  --color: var(--color-red-500);
  --color-active: var(--color-gray-100);
  --background: var(--color-red-100);
  --background-hover: var(--color-red-300);
  --background-active: var(--color-red-500);
  --disabled-line: var(--color-red-300);
}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {
  .c2 {
    width: 2.5rem;
  }
}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

<button
  aria-invalid="true"
  aria-pressed="false"
  class="c0 c1 c2"
  data-testid="box-0-0"
>
  <span>
    1
  </span>
</button>
`;

exports[`Sudoku component collision feedback should add error styles by collision on col 1`] = `
.c0 {
  display: grid;
  grid-auto-flow: column;
  gap: 0.25rem;
  place-items: center;
  min-height: 2rem;
  padding: 0 0.75rem;
  font-weight: 600;
  color: var(--color);
  text-transform: capitalize;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: var(--background);
  border-radius: 0.5rem;
  outline-color: var(--color);
  -webkit-transition: 100ms ease-out;
  transition: 100ms ease-out;
  -webkit-transition-property: background,color;
  transition-property: background,color;
  --color: var(--color-blue-400);
  --color-hover: var(--color);
  --color-active: var(--color-gray-100);
  --background: var(--color-blue-100);
  --background-hover: var(--color-blue-200);
  --background-active: var(--color-blue-400);
}

.c0:hover,
.c0:focus-visible {
  color: var(--color-hover);
  background: var(--background-hover);
}

.c0:active {
  color: var(--color-active);
  background: var(--background-active);
}

.c1[aria-pressed='true'] {
  --color: var(--color-gray-100);
  --background: var(--color-blue-400);
  --background-hover: var(--color-blue-500);
  --background-active: var(--color-blue-600);
}

.c1[aria-pressed='true']:focus {
  outline-color: var(--color-gray-900);
}

.c2 {
  position: relative;
  aspect-ratio: 1;
  display: grid;
  place-items: center;
  padding: 0;
  font-weight: 700;
  border-radius: 0.2rem;
  --color: var(--color-gray-900);
  --color-active: var(--color-gray-900);
  --background: var(--color-gray-200);
  --background-hover: var(--color-gray-300);
  --background-active: var(--color-gray-400);
  --disabled-line: var(--color-gray-300);
  --background: var(--color-gray-300);
  --background-hover: var(--color-gray-400);
  --background-active: var(--color-gray-500);
  --disabled-line: var(--color-gray-500);
}

.c2:disabled {
  background-image: linear-gradient( -45deg, var(--background) 15%, var(--disabled-line) 15% 25%, var(--background) 25% 35%, var(--disabled-line) 35% 45%, var(--background) 45% 55%, var(--disabled-line) 55% 65%, var(--background) 65% 75%, var(--disabled-line) 75% 85%, var(--background) 85% );
  --color: var(--color-gray-600);
}

.c2[aria-pressed='true'] {
  --color: var(--color-green-400);
  --color-active: var(--color-gray-100);
  --background: var(--color-green-100);
  --background-hover: var(--color-green-200);
  --background-active: var(--color-green-400);
  --disabled-line: var(--color-green-200);
}

.c2[aria-invalid='true'] {
  --color: var(--color-red-500);
  --color-active: var(--color-gray-100);
  --background: var(--color-red-100);
  --background-hover: var(--color-red-300);
  --background-active: var(--color-red-500);
  --disabled-line: var(--color-red-300);
}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {
  .c2 {
    width: 2.5rem;
  }
}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

<button
  aria-invalid="true"
  aria-pressed="false"
  class="c0 c1 c2"
  data-testid="box-0-0"
>
  <span>
    1
  </span>
</button>
`;

exports[`Sudoku component collision feedback should add error styles by collision on row 1`] = `
.c0 {
  display: grid;
  grid-auto-flow: column;
  gap: 0.25rem;
  place-items: center;
  min-height: 2rem;
  padding: 0 0.75rem;
  font-weight: 600;
  color: var(--color);
  text-transform: capitalize;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: var(--background);
  border-radius: 0.5rem;
  outline-color: var(--color);
  -webkit-transition: 100ms ease-out;
  transition: 100ms ease-out;
  -webkit-transition-property: background,color;
  transition-property: background,color;
  --color: var(--color-blue-400);
  --color-hover: var(--color);
  --color-active: var(--color-gray-100);
  --background: var(--color-blue-100);
  --background-hover: var(--color-blue-200);
  --background-active: var(--color-blue-400);
}

.c0:hover,
.c0:focus-visible {
  color: var(--color-hover);
  background: var(--background-hover);
}

.c0:active {
  color: var(--color-active);
  background: var(--background-active);
}

.c1[aria-pressed='true'] {
  --color: var(--color-gray-100);
  --background: var(--color-blue-400);
  --background-hover: var(--color-blue-500);
  --background-active: var(--color-blue-600);
}

.c1[aria-pressed='true']:focus {
  outline-color: var(--color-gray-900);
}

.c2 {
  position: relative;
  aspect-ratio: 1;
  display: grid;
  place-items: center;
  padding: 0;
  font-weight: 700;
  border-radius: 0.2rem;
  --color: var(--color-gray-900);
  --color-active: var(--color-gray-900);
  --background: var(--color-gray-200);
  --background-hover: var(--color-gray-300);
  --background-active: var(--color-gray-400);
  --disabled-line: var(--color-gray-300);
  --background: var(--color-gray-300);
  --background-hover: var(--color-gray-400);
  --background-active: var(--color-gray-500);
  --disabled-line: var(--color-gray-500);
}

.c2:disabled {
  background-image: linear-gradient( -45deg, var(--background) 15%, var(--disabled-line) 15% 25%, var(--background) 25% 35%, var(--disabled-line) 35% 45%, var(--background) 45% 55%, var(--disabled-line) 55% 65%, var(--background) 65% 75%, var(--disabled-line) 75% 85%, var(--background) 85% );
  --color: var(--color-gray-600);
}

.c2[aria-pressed='true'] {
  --color: var(--color-green-400);
  --color-active: var(--color-gray-100);
  --background: var(--color-green-100);
  --background-hover: var(--color-green-200);
  --background-active: var(--color-green-400);
  --disabled-line: var(--color-green-200);
}

.c2[aria-invalid='true'] {
  --color: var(--color-red-500);
  --color-active: var(--color-gray-100);
  --background: var(--color-red-100);
  --background-hover: var(--color-red-300);
  --background-active: var(--color-red-500);
  --disabled-line: var(--color-red-300);
}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {
  .c2 {
    width: 2.5rem;
  }
}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

<button
  aria-invalid="true"
  aria-pressed="false"
  class="c0 c1 c2"
  data-testid="box-0-0"
>
  <span>
    1
  </span>
</button>
`;

exports[`Sudoku component functionality with controls numpad should add selected styles to box with same value as selected numpad 1`] = `
.c0 {
  display: grid;
  grid-auto-flow: column;
  gap: 0.25rem;
  place-items: center;
  min-height: 2rem;
  padding: 0 0.75rem;
  font-weight: 600;
  color: var(--color);
  text-transform: capitalize;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: var(--background);
  border-radius: 0.5rem;
  outline-color: var(--color);
  -webkit-transition: 100ms ease-out;
  transition: 100ms ease-out;
  -webkit-transition-property: background,color;
  transition-property: background,color;
  --color: var(--color-blue-400);
  --color-hover: var(--color);
  --color-active: var(--color-gray-100);
  --background: var(--color-blue-100);
  --background-hover: var(--color-blue-200);
  --background-active: var(--color-blue-400);
}

.c0:hover,
.c0:focus-visible {
  color: var(--color-hover);
  background: var(--background-hover);
}

.c0:active {
  color: var(--color-active);
  background: var(--background-active);
}

.c1[aria-pressed='true'] {
  --color: var(--color-gray-100);
  --background: var(--color-blue-400);
  --background-hover: var(--color-blue-500);
  --background-active: var(--color-blue-600);
}

.c1[aria-pressed='true']:focus {
  outline-color: var(--color-gray-900);
}

.c2 {
  position: relative;
  aspect-ratio: 1;
  display: grid;
  place-items: center;
  padding: 0;
  font-weight: 700;
  border-radius: 0.2rem;
  --color: var(--color-gray-900);
  --color-active: var(--color-gray-900);
  --background: var(--color-gray-200);
  --background-hover: var(--color-gray-300);
  --background-active: var(--color-gray-400);
  --disabled-line: var(--color-gray-300);
  --background: var(--color-gray-300);
  --background-hover: var(--color-gray-400);
  --background-active: var(--color-gray-500);
  --disabled-line: var(--color-gray-500);
}

.c2:disabled {
  background-image: linear-gradient( -45deg, var(--background) 15%, var(--disabled-line) 15% 25%, var(--background) 25% 35%, var(--disabled-line) 35% 45%, var(--background) 45% 55%, var(--disabled-line) 55% 65%, var(--background) 65% 75%, var(--disabled-line) 75% 85%, var(--background) 85% );
  --color: var(--color-gray-600);
}

.c2[aria-pressed='true'] {
  --color: var(--color-green-400);
  --color-active: var(--color-gray-100);
  --background: var(--color-green-100);
  --background-hover: var(--color-green-200);
  --background-active: var(--color-green-400);
  --disabled-line: var(--color-green-200);
}

.c2[aria-invalid='true'] {
  --color: var(--color-red-500);
  --color-active: var(--color-gray-100);
  --background: var(--color-red-100);
  --background-hover: var(--color-red-300);
  --background-active: var(--color-red-500);
  --disabled-line: var(--color-red-300);
}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {
  .c2 {
    width: 2.5rem;
  }
}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

@media (min-width:768px) {

}

<button
  aria-invalid="false"
  aria-pressed="true"
  class="c0 c1 c2"
  data-testid="box-0-0"
>
  <span>
    1
  </span>
</button>
`;
